packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol

```diff
@@ -0,0 +1,264 @@
+            .DecreaseLiquidityParams({
+                tokenId: _params.tokenId,
+                liquidity: uint128(_params.liquidity.mul(_params.liquidityPercentage).div(1e18)),
+                amount0Min: _params.amount0Min,
+                amount1Min: _params.amount1Min,
+                deadline: block.timestamp
+            });
+        INonfungiblePositionManager(_nonfungiblePositionManager).decreaseLiquidity(decreaseParams);
+
+        uint256 wethAmount;
+        uint256 _wPowerPerpAmount;
+        (_isWethToken0)
+            ? (wethAmount, _wPowerPerpAmount) = INonfungiblePositionManager(_nonfungiblePositionManager).collect(
+                INonfungiblePositionManager.CollectParams({
+                    tokenId: _params.tokenId,
+                    recipient: address(this),
+                    amount0Max: type(uint128).max,
+                    amount1Max: type(uint128).max
+                })
+            )
+            : (_wPowerPerpAmount, wethAmount) = INonfungiblePositionManager(_nonfungiblePositionManager).collect(
+            INonfungiblePositionManager.CollectParams({
+                tokenId: _params.tokenId,
+                recipient: address(this),
+                amount0Max: type(uint128).max,
+                amount1Max: type(uint128).max
+            })
+        );
+
+        return (_wPowerPerpAmount, wethAmount);
+    }
+
+    /**
+     * @notice minth amount of wPowerPerp and LP in weth/wPowerPerp pool
+     * @param _controller wPowerPerp controller address
+     * @param _nonfungiblePositionManager Uni NonFungiblePositionManager address
+     * @param _wPowerPerp wPowerPerp contract address
+     * @param _wPowerPerpPool wPowerPerp Uni v3 pool
+     * @param _mintAndLpParams ControllerHelperDataType.MintAndLpParams struct
+     * @param _isWethToken0 bool variable indicate if Weth token is token0 in Uniswap v3 weth/wPowerPerp pool
+     * @return _vaultId and tokenId
+     */
+    function mintAndLp(address _controller, address _nonfungiblePositionManager, address _wPowerPerp, address _wPowerPerpPool, ControllerHelperDataType.MintAndLpParams calldata _mintAndLpParams, bool _isWethToken0) public returns (uint256, uint256) {
+        uint256 _vaultId = IController(_controller).mintWPowerPerpAmount{value: _mintAndLpParams.collateralToDeposit}(
+            _mintAndLpParams.vaultId,
+            _mintAndLpParams.wPowerPerpAmount,
+            0
+        );
+
+        // LP _mintAndLpParams._wPowerPerpAmount & _mintAndLpParams.collateralToLp in Uni v3
+        uint256 uniTokenId = lpWPowerPerpPool(
+            _controller,
+            _nonfungiblePositionManager,
+            _wPowerPerpPool,
+            _wPowerPerp,
+            _vaultId,
+            ControllerHelperDataType.LpWPowerPerpPool({
+                recipient: _mintAndLpParams.recipient,
+                ethAmount: _mintAndLpParams.collateralToLp,
```

Comment:
`_mintAndLpParams.collateralToLp` in case the WETH (wether amount0 or amount1) is sent by user as ETH
https://github.com/opynfinance/squeeth-monorepo/blob/feat/rebalance-no-vault/packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol#L195